testing/rclone/APKBUILD

```diff
@@ -0,0 +1,98 @@
+# Contributor: tcely <rclone+aports@tcely.33mail.com>
+# Maintainer: tcely <rclone+aports@tcely.33mail.com>
+_pkgname="rclone"
+pkgname="${_pkgname}"
+pkgver="1.48.0"
+pkgrel=0
+pkgdesc="rsync for cloud storage"
+url="https://rclone.org/"
+arch="all"
+license="MIT"
+options=""
+depends=""
+makedepends="bash git go perl"
+install=""
+subpackages=""
+
+gpg_signature_extensions="."
+gpgfingerprints="good:FBF7 37EC E9F8 AB18 604B  D2AC 9393 5E02 FF3B 54FA"
+gpgsource="${pkgname}-${pkgver}-sha256sums::https://downloads.rclone.org/v${pkgver}/SHA256SUMS"
+source="https://downloads.rclone.org/v${pkgver}/${_pkgname}-v${pkgver}.tar.gz $gpgsource"
+
+builddir="${srcdir}/${_pkgname}-v${pkgver}"
+export GOPATH="${srcdir}/go"
+
+sanitycheck() {
+	local source= sha512sums=
+	default_sanitycheck
+}
+
+cleanup_srcdir() {
+	go clean -modcache
+	default_cleanup_srcdir
+}
+
+build() {
+	make rclone
+}
+
+check() {
+	make quicktest
+}
+
+package() {
+	make DESTDIR="${pkgdir}" install
+}
+
+verify() {
```

Comment:
This is very strange.
Perhaps this functionality should wait until after abuild has these features? (I assume this is done to allow for gpg checks)

Reply:
This verifies the sources from the signed sha256sums provided by upstream. This allows for checking the full chain of trust before calculating sha512 sums.

Reply:
Again, I understand what this does.
I'm questioning why it's done here, rather than being added to abuild.
APKBUILDs should generally describe how to build and package package-specific things.
I'm not sure they're the correct place to prototype new abuild functionality.

Reply:
I'm classifying this as package specific because published sums vary in the way they are handled quite a bit.

Reply:
So should we go around every aport and add custom `verify()` handling to each of them, with no serious advantage to it?
The developer is perfectly capable of checking the validity of the hash on their own, without this large hacked-on package-specific function.

If we want to do it automatically, it should *still* be done (in a sufficiently generic way) within abuild.
Potentially, you could consider asking for a `default_verify` so that you can at least keep the original `verify()` behavior too, instead of hoping every "custom apkbuild verify" keeps up to date with upstream abuild!

Either way, I'm not seeing why this is absolutely necessary, especially considering all of the introduced problems.

Reply:
I absolutely want automated checks over manual work. I don't see how `default_verify` would help. I already did keep the original verify behavior.

Reply:
Consider other packages implementing this behavior.
Each of them will duplicate original `verify` behavior.
Then, imagine abuild's `verify()` is changed/improved/gets-new-functionality.
Then EVERY package with a custom `verify()` has to update the entire function.
As such, 100% of the common code SHOULD be in the common place (in this case, `abuild`).

In software development, this is normally referred to as the "DRY" (Don't Repeat Yourself) rule.

Ideally, however, `abuild` should be able to handle the entire scenario, not requiring any modification on the APKBUILD side (since the APKBUILD *should* just describe how to correctly build the package; verification of sources shouldn't really be a part of that).

Reply:
You seem to believe I duplicated common code for this function. That is not the case. I'm already using a call to the common code where possible. So, I still don't see how `default_verify` would help.

Reply:
Comparing the abuild.in version with yours (both cut out manually) shows a diff of 36 lines; 3 of which are context/separators, and one of which is a modified line (-1).
This results in an effective diff of 32 lines.
Your solution is 49 lines.
Therefore, you share a total of 17 lines exactly.
The abuild version is 17 lines in total.

This means that your content *ONLY* adds new code and copies the entire abuild.in function in verbatim before adding in new stuff.
I realize that your stuff is interleaved inside of the actual function, but the reality is that you *ARE* duplicating common code.

The correct solution was, and remains, to modify abuild to have support for this natively (whether fully, if possible, with some changes to how it's handled to be generic; or in a way that allows piggybacking on the main function in an extended way, which is what I meant by `default_verify`).
Even a standard `default_verify` would help, though it would require some changes in how you're currently handling this code.

Reply:
I think we disagree on what we mean by the words we are using. Modification is not the same as duplication. I'm not sure what you envision a useful purpose for `default_verify` would be.

Please provide a patch to `abuild` that handles this setup (and its variations) for verifying signed sums from upstream in a generic way rather than continuing to argue that my implementation is not package specific. If you have an idea for how `default_verify` would be useful, I'd be very happy to see it.

Reply:
There appears to be a misunderstanding here.
*You* are the one that wants this functionality merged.
I am perfectly ok with verifying my checksums on bumps, as I do for other packages (including ones that do not have any sort of gpg protection).
The onus on making an implementation and getting it merged is on *you* as a result.
As is, I have little interest in maintaining this monstrosity, and am simply proposing ways as to make it less cumbersome.

Your statement is making it sound like you have no intention on even attempting on patching abuild, and just want this one specific package (and/or others like it, which would have the issue mentioned above, whether or not you "agree with the language" used to describe it) to be some individually rolled prototypes.
That approach is inherently unmaintainable, as it scales exponentially in complexity as the number of handled packages increases.

If you're interest in feedback as to possible ways to handle varying upstream packages, do feel free to ask for that, but please do keep in mind that the one that's interested in having the functionality merged is *you*.

Reply:
> There appears to be a misunderstanding here.
> _You_ are the one that wants this functionality merged.
> I am perfectly ok with verifying my checksums on bumps, as I do for other packages (including ones that do not have any sort of gpg protection).
> The onus on making an implementation and getting it merged is on _you_ as a result.

I am not happy with "best guess" checksums when a better way of verifying the downloads is available. I made an implementation, you are the one complaining it is not good enough. In a fairly combative tone too.

> As is, I have little interest in maintaining this monstrosity, and am simply proposing ways as to make it less cumbersome.

No one has asked you to either participate in name calling or maintenance.

> 
> Your statement is making it sound like you have no intention on even attempting on patching abuild, and just want this one specific package (and/or others like it, which would have the issue mentioned above, whether or not you "agree with the language" used to describe it) to be some individually rolled prototypes.
> That approach is inherently unmaintainable, as it scales exponentially in complexity as the number of handled packages increases.
> 

You can find the patch to abuild here: https://github.com/alpinelinux/abuild/pull/57

I have no intention of trying to merge a package specific process into `abuild` and I don't agree with your assertion that doing so is the best way forward. If you have a better idea, you are welcome to attempt to get it merged into `abuild`.

> If you're interest in feedback as to possible ways to handle varying upstream packages, do feel free to ask for that, but please do keep in mind that the one that's interested in having the functionality merged is _you_.

That is not a true statement. I don't want to have this package specific verify routine merged into `abuild`. It is your previous posts that have left me with the understanding that you want me to attempt to merge all of the possible ways signed sums can be used by various packages into `abuild` for you so that you (and others) can specify a variable or some such in `APKBUILD` to use one of those options.

I'm simply not willing to do this work for you while my previous work sits unreviewed. I'm happily using a version of `abuild` that validates GPG signatures, so my issue is already solved.

Reply:
> I am not happy with "best guess" checksums when a better way of verifying the downloads is available.

How do you know I don't cross-check my checksums against upstream gpg?
Further, why do you care about how checksums are done for packages you don't maintain?
Certainly, the onus of making sure the package isn't broken or compromised is on me, and you need not worry about that.
Instead, I am recommending having a generic way to do what you want - have the better way of verifying the downloads available for all packages, including ones with this type of handling.

> I made an implementation, you are the one complaining it is not good enough. In a fairly combative tone too.

My tone is not meant to be combative, but exasperated.
You continue to defend stylistic choices (that go against common practices), even if I am to be the one maintaining the package (which I am perfectly willing to do and will do whether or not this eventually merges (assuming it merges in a way that is actually maintainable)), you try to make semantic arguments (that nothing is repeated from the original verify function, despite every single logic line being present, unmodified), and now appear to be arguing that my tone is wrong!
If I come across as combative, I'm sorry about that, but it is not intended to be that way, and would ask that you adjust accordingly.

> > As is, I have little interest in maintaining this monstrosity, and am simply proposing ways as to make it less cumbersome.
> 
> No one has asked you to either participate in name calling or maintenance.

As of currently, I maintain this package.
If it is merged in its current form, I will be the one that will have to maintain your code.
It is not up to spec (that you are reluctant to discuss), and has multiple other issues.
*I* will be the one that will need to look at every commit made to abuild in case `verify` is ever changed.
I do not wish to do that.
Further, I do also wish to have an up-to-date version of rclone available.
I suspect that this PR has been mostly untouched for a month not because rclone is controversial (see the PR that did get through), but because of the very large function and strange stylistic choices.
This suggests that any future bumps will cause similar delays.

> > Your statement is making it sound like you have no intention on even attempting on patching abuild, and just want this one specific package (and/or others like it, which would have the issue mentioned above, whether or not you "agree with the language" used to describe it) to be some individually rolled prototypes.
> > That approach is inherently unmaintainable, as it scales exponentially in complexity as the number of handled packages increases.
> 
> You can find the patch to abuild here: [alpinelinux/abuild#57](https://github.com/alpinelinux/abuild/pull/57)
> 
> I have no intention of trying to merge a package specific process into `abuild` and I don't agree with your assertion that doing so is the best way forward. If you have a better idea, you are welcome to attempt to get it merged into `abuild`.

This process is package specific because your original patch is not flexible enough.
I assert that it is possible to have a patch to `abuild` that *is* flexible enough, and only requires minimal changes on the APKBUILD side.
However, again, *I* have no interest in having generic (or non-generic) gpg verification processes in abuild.
I just want my packages to be up to date, and maintainable, so that I may keep them up to date, without, for instance, looking at every commit done to abuild since the last bump.

> > If you're interest in feedback as to possible ways to handle varying upstream packages, do feel free to ask for that, but please do keep in mind that the one that's interested in having the functionality merged is _you_.
> 
> That is not a true statement. I don't want to have this package specific verify routine merged into `abuild`. It is your previous posts that have left me with the understanding that you want me to attempt to merge all of the possible ways signed sums can be used by various packages into `abuild` for you so that you (and others) can specify a variable or some such in `APKBUILD` to use one of those options.

Again, I'm not talking about a package-specific verify routine.
I'm talking about a not-currently-written routine that will adequately handle your goal of allowing a full chain of trust within arbitrary packages.

Regarding what I want, it's simple.
I want packages to be up to date.
I want packages to be maintainable.
This changeset, as it is now, ultimately goes against both points (though only somewhat on one of the two counts).
I am *willing* (but not *interested*) in having a high quality generic approach to having chain of trust style verification.

> I'm simply not willing to do this work for you while my previous work sits unreviewed. I'm happily using a version of `abuild` that validates GPG signatures, so my issue is already solved.

Again, this would not be done *for me*.
I am perfectly content with this getting closed and you blaming me if the built version of rclone ends up compromised, when your approach would have caught the issue.

Since you appear to have trouble imagining a sufficiently generic algorithm for handling these cases, here is a potential sample:
1. Have `gpg_sources` - these must end in `.asc` or `.sig` (even if it's renamed using the `name::source` convention.
2. During fetching/verification, for every gpg_source, check if a non-.asc or non-.sig file exists in `sources`.
  If it does, verify that file using the `.asc|.sig` file.
  If it doesn't, consider it to be a clearsign file, and add the output to sources= (during runtime).

Past that, the only thing "special cases" such as this would need is an override of `abuild checksum`, to generate sums from the verified SHA*SUMS.txt file.

Reply:
> > I am not happy with "best guess" checksums when a better way of verifying the downloads is available.
> 
> How do you know I don't cross-check my checksums against upstream gpg?

I don't know that you do or don't. This is the problem. There really is not a lot that can be asserted by the sums matching. It is entirely possible for example that the original release was replaced with a trojan from the time the sums were checked into `aports.git` to the time the package was built.

> Further, why do you care about how checksums are done for packages you don't maintain?

As a user, I prefer to know (rather than assume) that the software that was built was that released by the authors.

> Certainly, the onus of making sure the package isn't broken or compromised is on me, and you need not worry about that.

I would call that unreasonable for both the user and maintainer roles. It's much better if both could be assured there is no compromise.

> Instead, I am recommending having a generic way to do what you want - have the better way of verifying the downloads available for all packages, including ones with this type of handling.
> 

Good recommendation. I don't have a generic way. When there is one or a solution that covers sufficient variations I'm happy to switch to it.

> > I made an implementation, you are the one complaining it is not good enough. In a fairly combative tone too.
> 
> My tone is not meant to be combative, but exasperated.
> You continue to defend stylistic choices (that go against common practices), even if I am to be the one maintaining the package (which I am perfectly willing to do and will do whether or not this eventually merges (assuming it merges in a way that is actually maintainable)), you try to make semantic arguments (that nothing is repeated from the original verify function, despite every single logic line being present, unmodified), and now appear to be arguing that my tone is wrong!
> If I come across as combative, I'm sorry about that, but it is not intended to be that way, and would ask that you adjust accordingly.
> 
I will adjust if you will. This is not a stylistic choice. I have an implementation that does what I wanted. You continue to say it is not good enough for you. Feel free to make it better. Until then, please stop telling me how you don't like it. Please also note that I did not say nothing is repeated. I said the function is not duplicated. Which means the verify function that exists in this pull request can not be replaced by the one in `abuild` while preserving the same operation.

> > > As is, I have little interest in maintaining this monstrosity, and am simply proposing ways as to make it less cumbersome.
> > 
> > 
> > No one has asked you to either participate in name calling or maintenance.
> 
> As of currently, I maintain this package.
> If it is merged in its current form, I will be the one that will have to maintain your code.

Yes, a situation that arose after your review of this pull request and the subsequent merging of your alternative pull request.

> It is not up to spec (that you are reluctant to discuss), and has multiple other issues.

I don't agree with that assessment.

> _I_ will be the one that will need to look at every commit made to abuild in case `verify` is ever changed.
> I do not wish to do that.

That's perfectly reasonable.

> Further, I do also wish to have an up-to-date version of rclone available.
> I suspect that this PR has been mostly untouched for a month not because rclone is controversial (see the PR that did get through), but because of the very large function and strange stylistic choices.
> This suggests that any future bumps will cause similar delays.

I don't know or care much about why the pull request was not interacted with. As I'm sure you also noticed, this project has not been promptly reviewing / acting on pull requests for quite a while now. The most common excuse in this period of time was the `3.10` release that was happening. Your suspicious or opinions about "strange stylistic choices" hold very little sway with me. Since I don't put much weight in these things I also don't believe your inference about future bumps is accurate.

> 
> > > Your statement is making it sound like you have no intention on even attempting on patching abuild, and just want this one specific package (and/or others like it, which would have the issue mentioned above, whether or not you "agree with the language" used to describe it) to be some individually rolled prototypes.
> > > That approach is inherently unmaintainable, as it scales exponentially in complexity as the number of handled packages increases.
> > 
> > 
> > You can find the patch to abuild here: [alpinelinux/abuild#57](https://github.com/alpinelinux/abuild/pull/57)
> > I have no intention of trying to merge a package specific process into `abuild` and I don't agree with your assertion that doing so is the best way forward. If you have a better idea, you are welcome to attempt to get it merged into `abuild`.
> 
> This process is package specific because your original patch is not flexible enough.
> I assert that it is possible to have a patch to `abuild` that _is_ flexible enough, and only requires minimal changes on the APKBUILD side.
> However, again, _I_ have no interest in having generic (or non-generic) gpg verification processes in abuild.
> I just want my packages to be up to date, and maintainable, so that I may keep them up to date, without, for instance, looking at every commit done to abuild since the last bump.
> 

You are making the assertion that it is possible. So the proof of the possible is left to you, as you also state that you are uninterested I understand that such proof is not forthcoming. However, please stop bemoaning my implementation that did not meet your notion of the perfect solution when you are unwilling to share anything better.

> > > If you're interest in feedback as to possible ways to handle varying upstream packages, do feel free to ask for that, but please do keep in mind that the one that's interested in having the functionality merged is _you_.
> > 
> > 
> > That is not a true statement. I don't want to have this package specific verify routine merged into `abuild`. It is your previous posts that have left me with the understanding that you want me to attempt to merge all of the possible ways signed sums can be used by various packages into `abuild` for you so that you (and others) can specify a variable or some such in `APKBUILD` to use one of those options.
> 
> Again, I'm not talking about a package-specific verify routine.
> I'm talking about a not-currently-written routine that will adequately handle your goal of allowing a full chain of trust within arbitrary packages.
> 

Again, please stop talking about the non-existent when reviewing the existing implementation. If you want to discuss your theoretical routine, you are welcome to propose such a thing in your own pull request.

> Regarding what I want, it's simple.
> I want packages to be up to date.
> I want packages to be maintainable.
> This changeset, as it is now, ultimately goes against both points (though only somewhat on one of the two counts).
> I am _willing_ (but not _interested_) in having a high quality generic approach to having chain of trust style verification.
> 
> > I'm simply not willing to do this work for you while my previous work sits unreviewed. I'm happily using a version of `abuild` that validates GPG signatures, so my issue is already solved.
> 
> Again, this would not be done _for me_.
> I am perfectly content with this getting closed and you blaming me if the built version of rclone ends up compromised, when your approach would have caught the issue.
> 

Then please stop asking for changes they are not coming. The current function met my needs.

> Since you appear to have trouble imagining a sufficiently generic algorithm for handling these cases, here is a potential sample:
> 
> 1. Have `gpg_sources` - these must end in `.asc` or `.sig` (even if it's renamed using the `name::source` convention.
> 2. During fetching/verification, for every gpg_source, check if a non-.asc or non-.sig file exists in `sources`.
>    If it does, verify that file using the `.asc|.sig` file.
>    If it doesn't, consider it to be a clearsign file, and add the output to sources= (during runtime).
> 
> Past that, the only thing "special cases" such as this would need is an override of `abuild checksum`, to generate sums from the verified SHA*SUMS.txt file.

This is too simplistic. You are welcome to implement it and see for yourself the issues you will run into.

Reply:
> I don't know that you do or don't. This is the problem. There really is not a lot that can be asserted by the sums matching. It is entirely possible for example that the original release was replaced with a trojan from the time the sums were checked into `aports.git` to the time the package was built.

If that happens, verification will fail.
Am I missing something?
Unless we assume that sha512 is broken, but if we assume that sha512 is broken, a trojan pretending to be rclone is the least of our concerns.

> > Further, why do you care about how checksums are done for packages you don't maintain?
> 
> As a user, I prefer to know (rather than assume) that the software that was built was that released by the authors.

Distributions add patches all the time.
If you want that (a legitimate desire, mind you!) I would recommend using something like flatpak when available.
Point taken, though.

> > Certainly, the onus of making sure the package isn't broken or compromised is on me, and you need not worry about that.
> 
> I would call that unreasonable for both the user and maintainer roles. It's much better if both could be assured there is no compromise.

I agree, I just don't want it to be done at the price that it is now.
As mentioned, I would strongly support a real generic way to do it in abuild.

> Good recommendation. I don't have a generic way. When there is one or a solution that covers sufficient variations I'm happy to switch to it.

> I will adjust if you will. This is not a stylistic choice. I have an implementation that does what I wanted. You continue to say it is not good enough for you. Feel free to make it better. Until then, please stop telling me how you don't like it. Please also note that I did not say nothing is repeated. I said the function is not duplicated. Which means the verify function that exists in this pull request can not be replaced by the one in `abuild` while preserving the same operation.

My concern is about needing to regularly check abuild commits in every custom-handled package to keep up with upstream.
What term is used (duplication, repetition, whatever) is irrelevant, imo.

> > It is not up to spec (that you are reluctant to discuss), and has multiple other issues.
> 
> I don't agree with that assessment.

- It actively avoids style guidelines (e.g empty variables)
- It does the _pkgname + pkgname thing to protect against an unlikely hypothetical
- It uses `{}`s in ways that will confuse both new people (since most packages don't use them) and experienced shell writers (since they will wonder why they are there)
- It actively introduces unneeded dependencies
- It actively deviates from standard practice regarding go (build system, GOPATH)
- It shuffles things around post-rebase (look at the current resulting diff)

A conversation can be had on whether or not everything should switch to a new spec (e.g GOPATH=$srcdir/go).
However, I don't think it can be argued that everything is "as it should be".

> > _I_ will be the one that will need to look at every commit made to abuild in case `verify` is ever changed.
> > I do not wish to do that.
> 
> That's perfectly reasonable.

Thank you.
I do believe that even if abuild isn't deeply patched, it should be possible to massage the function to a state where I wouldn't mind maintaining it.
However, based on further comments, I assume there is no intent on doing this.
Once your patch to abuild is merged, do feel free to poke me for brainstorming re: how to do it properly in a generic way.

> > Further, I do also wish to have an up-to-date version of rclone available.
> > I suspect that this PR has been mostly untouched for a month not because rclone is controversial (see the PR that did get through), but because of the very large function and strange stylistic choices.
> > This suggests that any future bumps will cause similar delays.
> 
> I don't know or care much about why the pull request was not interacted with. As I'm sure you also noticed, this project has not been promptly reviewing / acting on pull requests for quite a while now. The most common excuse in this period of time was the `3.10` release that was happening. Your suspicious or opinions about "strange stylistic choices" hold very little sway with me. Since I don't put much weight in these things I also don't believe your inference about future bumps is accurate.

I suppose we can disagree then.

> > > > Your statement is making it sound like you have no intention on even attempting on patching abuild, and just want this one specific package (and/or others like it, which would have the issue mentioned above, whether or not you "agree with the language" used to describe it) to be some individually rolled prototypes.
> > > > That approach is inherently unmaintainable, as it scales exponentially in complexity as the number of handled packages increases.
> > > 
> > > 
> > > You can find the patch to abuild here: [alpinelinux/abuild#57](https://github.com/alpinelinux/abuild/pull/57)
> > > I have no intention of trying to merge a package specific process into `abuild` and I don't agree with your assertion that doing so is the best way forward. If you have a better idea, you are welcome to attempt to get it merged into `abuild`.
> > 
> > 
> > This process is package specific because your original patch is not flexible enough.
> > I assert that it is possible to have a patch to `abuild` that _is_ flexible enough, and only requires minimal changes on the APKBUILD side.
> > However, again, _I_ have no interest in having generic (or non-generic) gpg verification processes in abuild.
> > I just want my packages to be up to date, and maintainable, so that I may keep them up to date, without, for instance, looking at every commit done to abuild since the last bump.
> 
> You are making the assertion that it is possible. So the proof of the possible is left to you, as you also state that you are uninterested I understand that such proof is not forthcoming. However, please stop bemoaning my implementation that did not meet your notion of the perfect solution when you are unwilling to share anything better.

I bemoan your solution because I see no way out for myself.
I just want my packages to be up to date, and usually that means doing it myself (my current mean time to update is under 24h, though I expect it could increase slightly as the number of packages increases).

> Again, please stop talking about the non-existent when reviewing the existing implementation. If you want to discuss your theoretical routine, you are welcome to propose such a thing in your own pull request.

Ok fair.
Regarding the current implementation - I do not want to maintain it as-is.
I do want to maintain the package to make sure that it is consistently up to date, and will put in effort in that direction.
I also don't necessarily trust other people to keep things fully up to date with the same level of zest as me.

re: theoretical proposal, I will open an issue after the move to gitlab, and potentially take a closer look.

> > 1. Have `gpg_sources` - these must end in `.asc` or `.sig` (even if it's renamed using the `name::source` convention.
> > 2. During fetching/verification, for every gpg_source, check if a non-.asc or non-.sig file exists in `sources`.
> >    If it does, verify that file using the `.asc|.sig` file.
> >    If it doesn't, consider it to be a clearsign file, and add the output to sources= (during runtime).
> > 
> > Past that, the only thing "special cases" such as this would need is an override of `abuild checksum`, to generate sums from the verified SHA*SUMS.txt file.
> 
> This is too simplistic. You are welcome to implement it and see for yourself the issues you will run into.

Can you elaborate?
If this does have fundamental problems, I'd rather not spend the time potentially implementing it.

Reply:
> > I don't know that you do or don't. This is the problem. There really is not a lot that can be asserted by the sums matching. It is entirely possible for example that the original release was replaced with a trojan from the time the sums were committed into `aports.git` to the time the package was built.
> 
> If that happens, verification will fail.
> Am I missing something?
> Unless we assume that sha512 is broken, but if we assume that sha512 is broken, a trojan pretending to be rclone is the least of our concerns.
> 

Why would verification fail? When the sums that were created by `abuild` were those of the trojan, they will match just fine. You have nothing to assert that the upstream release is the same as what the authors intended after all.

> > > It is not up to spec (that you are reluctant to discuss), and has multiple other issues.
> > 
> > 
> > I don't agree with that assessment.
> 
> * It actively avoids style guidelines (e.g empty variables)

Removing empty variables or in other words removing the template is not helpful for new users and serves no good purpose. If this is out of spec, I'd rather see the spec change. I haven't seen any spec around this yet at all.

> * It does the _pkgname + pkgname thing to protect against an unlikely hypothetical

It does indeed do this, I don't consider that an issue or out of spec.

> * It uses `{}`s in ways that will confuse both new people (since most packages don't use them) and experienced shell writers (since they will wonder why they are there)

What most packages do is immaterial. You can find the explanation for why always quoting and bracing variable uses is better at: https://gist.github.com/tcely/8ef31809f04a494a27ad79d49afdf167.

> * It actively introduces unneeded dependencies

This statement is only true if you ignore the upstream `Makefile`.

> * It actively deviates from standard practice regarding go (build system, GOPATH)

Certainly the `rclone` `Makefile` is not standard. However, as we already covered, using the build system that upstream provided is hardly against standard practice. Please address the `rclone` project to help them improve the `Makefile` instead of ignoring it.

> * It shuffles things around post-rebase (look at the current resulting diff)
> 

The only change in ordering I found was `options` which is not included in the [template](https://github.com/alpinelinux/abuild/blob/297de93aef3110f6ced8926e64e4204692b3b74e/newapkbuild.in#L241-L257) and is found all over the place as a result. In fact, I took specific  care to move my functions around to match the already committed order.

> A conversation can be had on whether or not everything should switch to a new spec (e.g GOPATH=$srcdir/go).
> However, I don't think it can be argued that everything is "as it should be".
> 

I certainly agree with that. `GOPATH` should really be treated much more like `pkgbasedir` in my opinion. That likely requires both `setup_go` and `cleanup_go` functions.

> > > 1. Have `gpg_sources` - these must end in `.asc` or `.sig` (even if it's renamed using the `name::source` convention.
> > > 2. During fetching/verification, for every gpg_source, check if a non-.asc or non-.sig file exists in `sources`.
> > >    If it does, verify that file using the `.asc|.sig` file.
> > >    If it doesn't, consider it to be a clearsign file, and add the output to sources= (during runtime).
> > > 
> > > Past that, the only thing "special cases" such as this would need is an override of `abuild checksum`, to generate sums from the verified SHA*SUMS.txt file.
> > 
> > 
> > This is too simplistic. You are welcome to implement it and see for yourself the issues you will run into.
> 
> Can you elaborate?
> If this does have fundamental problems, I'd rather not spend the time potentially implementing it.

Handling clear signatures, encrypted signatures, and detached signatures is a good first step. After that there's not a good standard about what the sums file will contain. Some considerations are input format (there are at least two common ones), which files are summed, and which algorithm is used.

As for `abuild` modifications, you need to pass the sanity checks, then you need fetch to run to gather all the remote files, then you need verify to pass, then you can think about generating the new sums list for everything that's in sources. I don't see a need to use only the sums that were signed, but that is certainly a valid option.